File: Yandex/Practicum/Algorightms_and_Data_Structures/5_sprint/E._Дерево_поиска/solution.py
import os

LOCAL = os.environ.get('REMOTE_JUDGE', 'false') != 'true'

if LOCAL:
    class Node:
        def __init__(self, value, left=None, right=None):
            self.value = value
            self.right = right
            self.left = left


def solution(root, min_value: int = float('-inf'), max_value: int = float('inf')) -> bool:
    """
    Time Complexity O(N)
    Memory Complexity O(H)
    """
    if not root:
        return True
    elif root.value <= min_value or root.value >= max_value:
        return False
    return solution(root.left, min_value, root.value) and solution(root.right, root.value, max_value)


def test():
    node1 = Node(1, None, None)
    node2 = Node(4, None, None)
    node3 = Node(3, node1, node2)
    node4 = Node(8, None, None)
    node5 = Node(5, node3, node4)

    assert solution(node5)
    node2.value = 5
    assert not solution(node5)


if __name__ == '__main__':
    test()

File: Yandex/Practicum/Algorightms_and_Data_Structures/5_sprint/E._Дерево_поиска/solution.cpp
#ifndef REMOTE_JUDGE
struct Node {
  int value;
  const Node* left = nullptr;
  const Node* right = nullptr;
  Node(int value, Node* left, Node* right) : value(value), left(left), right(right) {}
};
#endif

#ifdef REMOTE_JUDGE
#include "solution_tree.h"
#endif
#include <cmath>
#include <iostream>
#include <cassert>
#include <cstdint>

using namespace std;

bool dfs(const Node* root, int minValue = INT32_MIN, int maxValue = INT32_MAX) {
    if (!root) {
        return true;
    } else if (root->value <= minValue || root->value >= maxValue) {
        return false;
    }
    return dfs(root->left, minValue, root->value) && dfs(root->right, root->value, maxValue);
};

bool Solution(const Node* root) {
    /*
    Time Complexity O(N)
    Memory Complexity O(H)
    */
    return dfs(root);
}


#ifndef REMOTE_JUDGE
void test() {
    {
      Node node1({1, nullptr, nullptr});
      Node node2({4, nullptr, nullptr});
      Node node3({3, &node1, &node2});
      Node node4({8, nullptr, nullptr});
      Node node5({5, &node3, &node4});
      assert(Solution(&node5));
    }
    {
      Node node1({1, nullptr, nullptr});
      Node node2({5, nullptr, nullptr});
      Node node3({3, &node1, &node2});
      Node node4({8, nullptr, nullptr});
      Node node5({5, &node3, &node4});
      assert(!Solution(&node5));
    }
}


int main() {
  test();
}
#endif